File: Data Structures & Algorithms/max-water-container/submission-7.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l, r  = 0, len(heights)-1
        maxarea = 0

        while l < r:
            y = min(heights[l], heights[r])
            x = r - l
            maxarea = max(maxarea, x * y)

            if heights[l] > heights[r]:
                r -= 1
            else:
                l += 1
        
        return maxarea


        

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-7.java
class Solution {
    static class Bar
    {
        int start;
        int height;
        Bar(int start, int height)
        {
            this.start = start;
            this.height = height;
        }
    }
    public int largestRectangleArea(int[] heights) {
        Deque<Bar> stack = new ArrayDeque<>();// each element = [index, height]
        int maxArea = 0;

        for (int i = 0; i < heights.length; i++) {
            int start = i;

            // while current bar is shorter than the one on stack top
            // pop from stack and calculate area
            while (!stack.isEmpty() && stack.peek().height > heights[i]) {
                Bar top = stack.pop();
                maxArea = Math.max(maxArea, top.height * (i - top.start));
                start = top.start; // we can extend the new bar to this popped index
            }

            stack.push(new Bar(start, heights[i]));
        }

        // handle bars still in stack
        while (!stack.isEmpty())
        {
            Bar pair = stack.pop();
            maxArea = Math.max(maxArea, pair.height * (heights.length - pair.start));
        }

        return maxArea;
    }
}
